feat: Integration of `correction-lib` within CorrectedJets/METFactory

anpicci opened the pull request:
This PR is intended to track the progress on integrating `correction-lib` within the coffea JME tools.

The very first contribution is the calculation of the JECs using inputs from the `correction-lib` tool.

FYI @lgray @nsmith-

anpicci committed:
First implementation of correction-lib for JEC

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -98,8 +98,21 @@ def getfunction(layout, depth):
+
 class CorrectedJetsFactory(object):
-    def __init__(self, name_map, jec_stack):
+    def __init__(self, name_map, jec_stack, jec_names, jec_year):
```
Please do not alter the signature of this function (especially with new non-defaulted arguments). We want to maintain backwards compatibility for the time being. 

Could you try implementing this such that the code detects correction-lib inputs coming in from the JEC stack, and then follow a different code path in the case of correctionlib inputs?

anpicci replied:
@lgray thank you for pointing out this. Would it work if `jec_stack` is a list or a dict if one wants to use `correction-lib`? IMU, the JECStack class (as well as the auxiliary ones) is not needed when using `correction-lib`, as the latter already has functionalities to get the right numbers from the formulas that are part of the JME corrections/

lgray replied:
The JEC stack also defines the mapping from names in your data to the arguments of the functions to evaluate. Sure, you can have the `jec_stack` be something much simpler (like a list) if correctionlib does a bunch of the bookkeeping for you.

Though I think you'll need to do some arranging yourself for the JES uncertainties.

lgray commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -171,6 +191,29 @@ def build(self, jets, lazy_cache):
         # take care of nominal JEC (no JER if available)
+        ## General setup to use correction-lib
+        clib_json = self.jec_year
+        json_path = f"/cvmfs/cms.cern.ch/rsync/cms-nanoAOD/jsonpog-integration/POG/JME/{clib_json}/jet_jerc.json.gz"
```
we do not put hard dependencies on cvmfs paths in coffea.

lgray replied:
This should be configurable.

I'd also be careful of building new objects every time you call `uncertainties`, this can become very slow.

garvitaa replied:
Also, the json file name would change depending on if it's AK8 or AK4. So, I would also recommend against hard-coding this and instead keeping this in the jec_stack.

garvitaa commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -140,7 +156,11 @@ def __init__(self, name_map, jec_stack):
+        self.jec_names = jec_names
+        self.jec_year = jec_year
+        ## the last element of jec_names has to be a boolean, specifying if the use wants to save the different correction levels
+        self.separated = self.jec_names.pop()
```
This could be a useful debug feature, but from the JME point of view, providing just the name of the JEC tag for the required step should suffice. I don't foresee a need for analyses to store the result of each step, but I'm open to hearing about any exceptions.

anpicci replied:
@garvitaa I would suggest keeping it to provide users more flexibility, without the need for them to put the hands on the coffee scripts

anpicci committed:
Updated version of CorrectedJetsFactory

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

anpicci commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -182,17 +259,85 @@ def build(self, jets, lazy_cache):
+
+                cumCorr = None
+                if len(corrections_list) > 0:
+                    ones = numpy.ones_like(corrections_list[-1], dtype=numpy.float32)
```
Try without `ones`

anpicci commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -220,28 +365,80 @@ def build(self, jets, lazy_cache):
+            if self.jec_stack.jer is not None and self.jec_stack.jersf is not None:
+                has_jer = True
+        elif self.tool == "clib":
             has_jer = True
```
Add a sanity check if the input contains jer or not

anpicci commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -220,28 +365,80 @@ def build(self, jets, lazy_cache):
+            elif self.tool == "clib":
+                ## needed to attach to jets the JECs
+                jer_out_parms = out.layout.parameters
+                jer_out_parms["corrected"] = True
```
Consider to add a method here:
https://github.com/CoffeaTeam/coffea/blob/master/src/coffea/nanoevents/methods/nanoaod.py#L499

anpicci commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -313,12 +522,24 @@ def build(self, jets, lazy_cache):
+                        out_dict["jet_energy_resolution"], numpy.float32
+                    ),
+                    awkward.values_astype(
+                        out_dict["jet_resolution_rand_gauss"], numpy.float32
```
Try to use the `correction-lib` random values

anpicci committed:
Merge pull request #1 from scikit-hep/backports-v0.7.x

Backports v0.7.x

anpicci committed:
new implementation with dataclass

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lgray committed:
use coffea's JECStack

lgray commented on coffea/jetmet_tools/CorrectedJetsFactory.py:
```diff
@@ -114,53 +150,65 @@ def __init__(self, name_map, jec_stack):
-        self.name_map = name_map
-        self.jec_stack = jec_stack
-
-    def uncertainties(self):
```
why was this function removed?

Justify and alter tests or please add it back!

anpicci committed:
merging upstream backport-0.7.x

anpicci committed:
Merging with the upstream backports-v0.7.x branch

feat:Merging with the upstream backports-v0.7.x branch

anpicci committed:
Add TaskVine executor, modernize CI/packaging, and refresh 0.7.x docs links (#14)

Add TaskVine executor + CI test; modernize packaging/deps; fix jet/MET + SciPy compat; update docs links

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci
